tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+    }
+}
+
+impl CalyxWriter for Control {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        self.attributes.write(f)?;
+        match &self.value {
+            ControlValue::Empty => {}
+            ControlValue::Enable(group) => {
+                write!(f, "{};", group.borrow().name)?;
+            }
+            ControlValue::Seq(body) => {
+                writeln!(f, "seq {{")?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::Par(body) => {
+                writeln!(f, "par {{")?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::While(cond, group, body) => {
+                writeln!(
+                    f,
+                    "while {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+            ControlValue::If(cond, group, body_true, body_false) => {
+                writeln!(
+                    f,
+                    "if {}{} {{",
+                    cond,
+                    if let Some(group) = group {
+                        format!(" with {}", group.borrow().name)
+                    } else {
+                        "".into()
+                    }
+                )?;
+                f.push();
+                for node in body_true {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                writeln!(f, "}} else {{")?;
+                f.push();
+                for node in body_false {
+                    node.writeln(f)?;
+                }
+                f.pop();
+                write!(f, "}}")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// See `calyx_ir::Component`.
+pub struct Component {
+    attributes: Attributes,
+    is_comb: bool,
+    name: String,
+    inputs: Vec<Port>,
+    outputs: Vec<Port>,
+    cells: Vec<RRC<Cell>>,
+    groups: Vec<RRC<Group>>,
+    control: Control,
+    continuous_assignments: Vec<Assignment>,
+}
+
+impl Component {
+    fn new<S: ToString>(name: S, is_comb: bool) -> Self {
+        Self {
+            attributes: vec![],
+            is_comb,
+            name: name.to_string(),
+            inputs: vec![
+                Port::new_in_comp("go", 1)
+                    .with_attribute(Attribute::bool("go")),
```

Comment:
nit: technically `go` is not a `BoolAttr` but I'm sure this is fine

Reply:
I can have boolattrs just display the `(1)` after them and this should work in general?